processors/zipkin/src/main/java/org/hawkular/apm/processor/zipkin/CompletionTimeProcessingDeriver.java

```diff
@@ -0,0 +1,147 @@
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.hawkular.apm.processor.zipkin;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import org.hawkular.apm.api.model.Property;
+import org.hawkular.apm.api.model.events.CompletionTime;
+import org.hawkular.apm.server.api.model.zipkin.Annotation;
+import org.hawkular.apm.server.api.model.zipkin.Span;
+import org.hawkular.apm.server.api.services.SpanCache;
+import org.hawkular.apm.server.api.task.AbstractProcessor;
+import org.hawkular.apm.server.api.task.RetryAttemptException;
+
+/**
+ * Processing deriver of completion time of trace reported by zipkin instrumentation.
+ *
+ * @author Pavol Loffay
+ */
+@Singleton
+public class CompletionTimeProcessingDeriver extends AbstractProcessor<CompletionTimeProcessing, CompletionTimeProcessing>{
+
+    private SpanCache spanCache;
+
+
+    public CompletionTimeProcessingDeriver() {
+        super(ProcessorType.OneToOne);
+    }
+
+    @Inject
+    public CompletionTimeProcessingDeriver(SpanCache spanCache) {
+        this();
+        this.spanCache = spanCache;
+    }
+
+
+    @Override
+    public CompletionTimeProcessing processOneToOne(String tenantId, CompletionTimeProcessing completionTimeProcessing)
+            throws RetryAttemptException{
+
+        Span rootSpan = completionTimeProcessing.getRootSpan();
+
+        List<Span> trace = spanCache.getTrace(tenantId, rootSpan.getTraceId());
+        if (trace == null) {
+            throw new RetryAttemptException("trace not present in the cache");
+        }
+
+        Annotation lastAnnotation = extractLastAnnotation(trace);
+        if (lastAnnotation == null) {
+            return null;
+        }
+
+        /**
+         * TODO HWKAPM-348
+         *
+         * When there is new information recorded in trace (annotation with bigger timestamp)
+         * then wait for a fixed amount of time {@link #getRetryDelay(List, int)} to calculate
+         * trace completion time for all reported spans (async span can be recorded aby time after root
+         * span has been recorded). This approach does not deal with long run async spans,
+         * because it waits only for a fixed amount of time.
+         */
+        if (completionTimeProcessing.getLastTimestamp() == null ||
```

Comment:
If I'm understanding the logic correctly, the processing will continue only if a new span has been added (with greater timestamp) since the last processing was performed?

At the moment you don't override the getDeliveryDelay method, so it will be processed immediately - so probably should set a value to ensure the processing isn't performed too quickly, and therefore miss subsequent spans in the same trace.